### PR TITLE
[GO-000] Locker interface with drivers

### PR DESCRIPTION
**Pull request by im-kulikov**

### CODE REVIEW

File: locker/redis/options.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	for _, op := range opts {
+		op(o)
+	}
+}
+
+// Client is a minimal client interface.
+type Client = lock.RedisClient
+
+// Timeout closure
+func Timeout(v interface{}) locker.Option {
```
Comment (jenchik) [inline locker/redis/options.go]: А зачем здесь interface{} ? Можно же сразу тут time.Duration

File: locker/redis/options.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		if !ok {
+			return errors.New("bad option timeout, must be time.Duration")
+		}
+		o.(*Options).Timeout = val
+		return nil
+	}
+}
+
+// Count closure to set RetryCount
+func Count(v interface{}) locker.Option {
```
Comment (jenchik) [inline locker/redis/options.go]: ```Count(v int)```

File: locker/redis/options.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		if !ok {
+			return errors.New("bad option count, must be int")
+		}
+		o.(*Options).RetryCount = val
+		return nil
+	}
+}
+
+// Delay closure to set RetryDelay
+func Delay(v interface{}) locker.Option {
```
Comment (jenchik) [inline locker/redis/options.go]: mb ```Delay(v time.Duration)``` ?

File: locker/redis/options.go (hunk 1 above)
Comment (im-kulikov) [reply locker/redis/options.go]: nop..

File: locker/redis/options.go (hunk 2 above)
Comment (im-kulikov) [reply locker/redis/options.go]: nop

File: locker/redis/options.go (hunk 3 above)
Comment (im-kulikov) [reply locker/redis/options.go]: nop

**Comment by im-kulikov** [comment]
@jenchik hm.. I'll checked my code... you'll right.. fixed

File: locker/locker.go [hunk 4]
```diff
@@ -1,44 +1,14 @@
...
-func (l *Lock) Run(key string, timeout time.Duration, handler func()) {
-	opts := &lock.Options{
-		RetryCount:  l.locker.RetryCount,
-		RetryDelay:  l.locker.RetryDelay,
-		LockTimeout: timeout,
-	}
-	if err := lock.Run(l.redis, key, opts, handler); err != nil {
-		log.Errorf("Locker error: %v", err.Error())
-	}
+	Run(key string, handler func(), options ...Option)
```
Comment (jenchik) [inline locker/locker.go]: mb need returns error?

Comment (jenchik) [reply locker/locker.go]: We do not know which option is valid. The driver can only check at runtime. I cannot say how better

Comment (im-kulikov) [reply locker/locker.go]: fixed

File: locker/redis/redis.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	return &Lock{
+		redis: redis,
+	}
+}
+
+// Run runs a callback handler with a Redis lock.
+func (l *Lock) Run(key string, handler func(), options ...locker.Option) error {
+	opts := new(Options)
+
+	if err := opts.Parse(options...); err != nil {
```
Comment (jenchik) [inline locker/redis/redis.go:26]: mb?
```
return wrap.WithMessage(err, "locker parse")
```

File: locker/redis/redis.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		log.Debugf("Locker parse: %v", err)
+		return err
+	}
+
+	if err := lock.Run(l.redis, key, &lock.Options{
+		LockTimeout: opts.Timeout,
+		RetryCount:  opts.RetryCount,
+		RetryDelay:  opts.RetryDelay,
+	}, handler); err != nil {
```
Comment (jenchik) [inline locker/redis/redis.go:35]: mb?
```
return wrap.WithMessage(err, "locker error")
```

File: locker/locker.go [hunk 7]
```diff
@@ -1,44 +1,14 @@
...
+}
 
-	"github.com/bsm/redis-lock"
-	"github.com/cryptopay-dev/yaga/logger/log"
-	"github.com/go-redis/redis"
-)
+// Option closure
+type Option = func(opt Options) error
 
 // Locker interface to abstract bsm/redis-lock
```
Comment (jenchik) [inline locker/locker.go]: Comment is not true

File: locker/redis/redis.go (hunk 5 above)
Comment (im-kulikov) [reply locker/redis/redis.go:26]: fixed

File: locker/redis/redis.go (hunk 6 above)
Comment (im-kulikov) [reply locker/redis/redis.go:35]: fixed

**Review by pcherednichenko — APPROVED** [review]

**Review by chapsuk — APPROVED** [review]

File: locker/redis/redis.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+// New redis locker
+func New(redis Client) locker.Locker {
+	return &Lock{
+		redis: redis,
+	}
+}
+
+// Run runs a callback handler with a Redis lock.
+func (l *Lock) Run(key string, handler func(), options ...locker.Option) error {
```
Comment (jenchik) [inline locker/redis/redis.go:23]: mb?
```
	if cfg, found := l.cfgmap[key]; !found {
		opts := new(Options)
		if err := opts.Parse(options...); err != nil {
			// ...
		}
		// ...
		cfg := &lock.Options{
			LockTimeout: opts.Timeout,
			RetryCount:  opts.RetryCount,
			RetryDelay:  opts.RetryDelay,
		}
		l.cfgmap[key] = cfg
	}
	if err := lock.Run(l.redis, key, cfg, handler); err != {
		// ...
	}
```

**Review by jenchik — APPROVED** [review]

**Pull request merged**